resotocore/resotocore/infra_apps/package_manager.py

```diff
@@ -0,0 +1,229 @@
+    git_url: str
+
+
+InstallationSource = Union[FromHttp, FromGit]
+
+
+@frozen
+class InfraAppPackage:
+    manifest: AppManifest
+    source: InstallationSource
+
+
+PackageEntityDb = EntityDb[InfraAppName, InfraAppPackage]
+
+
+def app_manifest_entity_db(db: AsyncArangoDB, collection: str) -> ArangoEntityDb[InfraAppName, InfraAppPackage]:
+    return ArangoEntityDb(db, collection, InfraAppPackage, lambda k: InfraAppName(k.manifest.name))
+
+
+def config_id(name: InfraAppName) -> ConfigId:
+    return ConfigId(f"resoto/apps/{name}")
+
+
+class PackageManager:
+    def __init__(
+        self,
+        entity_db: PackageEntityDb,
+        config_handler: ConfigHandler,
+        repos_directory: Path = Path("/tmp/resoto-package-manager-repos"),
+        check_interval: timedelta = timedelta(hours=1),
+        cleanup_after: timedelta = timedelta(days=1),
+        current_epoch_seconds: Callable[[], float] = time.time,
+    ) -> None:
+        """
+        Package manager for infra apps.
+
+        Args:
+            entity_db: EntityDb for storing app manifests
+            config_handler: ConfigHandler for storing app configs
+            repos_directory: Directory where repos are temporarily cloned
+            check_interval: Interval between checks for old repos cleanup
+            cleanup_after: Time after which old repos are deleted from cache
+        """
+
+        self.entity_db = entity_db
+        self.config_handler = config_handler
+        self.install_delete_lock: Optional[Lock] = None
+        self.update_lock: Optional[Lock] = None
+        self.update_all_lock: Optional[Lock] = None
+        self.repos_directory: Path = repos_directory
+        self.check_interval = check_interval
+        self.cleanup_after = cleanup_after
+        self.current_epoch_seconds = current_epoch_seconds
+        self.cleanup_task: Optional[asyncio.Task[None]] = None
+
+    async def start(self) -> None:
+        self.install_delete_lock = asyncio.Lock()
+        self.update_lock = asyncio.Lock()
+        self.update_all_lock = asyncio.Lock()
+        self.cleanup_task = asyncio.create_task(self._cleanup_old_repos())
+
+    async def stop(self) -> None:
+        if self.cleanup_task:
+            self.cleanup_task.cancel()
+            await self.cleanup_task
+
+    def list(self) -> AsyncGenerator[InfraAppName, None]:
+        return self.entity_db.keys()
+
+    async def info(self, name: InfraAppName) -> Optional[AppManifest]:
+        if package := await self.entity_db.get(name):
+            return package.manifest
+        return None
+
+    async def update(self, name: InfraAppName) -> Optional[AppManifest]:
+        if not self.update_lock:
+            raise RuntimeError("PackageManager not started")
+
+        async with self.update_lock:
+            if package := await self.entity_db.get(name):
+                await self.delete(package.manifest.name)
+                return await self.install(package.manifest.name, package.source)
```

Comment:
Let's divide it into:
- get app manifest

check the manifest version - only update if required. If yes than:
- delete manifest
- insert manifest